controller_manager/test/test_control_loop.cpp

```diff
@@ -0,0 +1,194 @@
+}
+
+TEST(TestControlLoop, DurationIsPeriod)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 10;
+  };
+
+  // GIVEN do_work function that records duration values
+  std::vector<rclcpp::Duration> durations;
+  auto do_work = [&durations](rclcpp::Time, rclcpp::Duration duration) {
+    durations.push_back(duration);
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the average duration to be close to the period
+  // accurate to 0.1ms. The period we set was 10ms.
+  auto const average_duration = std::chrono::nanoseconds{
+    std::reduce(durations.cbegin() + 1, durations.cend(), rclcpp::Duration{0, 0}).nanoseconds() /
+    (durations.size() - 1)};
+  EXPECT_NEAR(average_duration.count(), 10'000'000, 100'000);
+}
+
+TEST(TestControlLoop, FirstDurationIsSmall)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 2;
+  };
+
+  // GIVEN do_work function that records duration values
+  std::vector<rclcpp::Duration> durations;
+  auto do_work = [&durations](rclcpp::Time, rclcpp::Duration duration) {
+    durations.push_back(duration);
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the first duration to be near 0
+  EXPECT_NEAR(durations.at(0).nanoseconds(), 0, 1'000);
+}
+
+TEST(TestControlLoop, DoWorkDelayInDuration)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 3;
+  };
+
+  // GIVEN do_work function takes longer than the period
+  std::vector<rclcpp::Duration> durations;
+  auto do_work = [&durations](rclcpp::Time, rclcpp::Duration duration) {
+    durations.push_back(duration);
+    std::this_thread::sleep_for(20'000'000ns);
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the second duration to be near the time the work function took
+  EXPECT_NEAR(durations.at(1).nanoseconds(), 20'000'000, 1'000'000);
+}
+
+TEST(TestControlLoop, LoopTimeBeforeWorkTime)
+{
+  auto loop_iterations = 0;
+  auto ok = [&loop_iterations]() {
+    loop_iterations++;
+    return loop_iterations < 3;
+  };
+
+  // GIVEN do_work function takes longer than the period
+  std::vector<rclcpp::Time> loop_times;
+  std::vector<rclcpp::Time> work_times;
+  auto do_work = [&loop_times, &work_times](rclcpp::Time time, rclcpp::Duration) {
+    loop_times.push_back(time);
+    work_times.push_back(now());
+  };
+
+  // WHEN we run the ControlLoop
+  ControlLoop(10'000'000ns, now, ok, do_work, sleep_until);
+
+  // THEN we expect the work times to be after the loop times
+  EXPECT_GT(work_times.at(0), loop_times.at(0));
+  EXPECT_GT(work_times.at(1), loop_times.at(1));
+}
+
+TEST(TestControlLoop, LoopTimeSeparatedByPeriod)
```

Comment:
How is this different than `DurationIsPeriod` test?
- manually calculates duration
- checks 1 loop rather than averaging over 10

bit of a nit, but my pref would to put those checks in the DurationIsPeriod test

Reply:
'LoopTimeCloseToReportedDuration' seems to do this, but more explicitly

Reply:
```suggestion
TEST(TestControlLoop, ReportedTimesMatchPeriod)
```
I see the differences now and made some naming suggestions from this fresh set of eyes